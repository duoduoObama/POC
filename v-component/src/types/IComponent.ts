import { type } from "os";

export interface IComponent {
    model?: ISchema,
}
export interface ISchema {
    get id(): string;
    get componentName(): string;
    get type(): string;
    get text(): string;
    get group(): string[];
    get createTime(): Date;
    get image(): string;
    get initStyle(): string;
    set initStyle(value: string);
    get description(): string;
    get options(): any;
    get schema(): ICustomSchema;
    get onMessageMeta(): IMessageMeta;
    set onMessageMeta(value: IMessageMeta);
    get onDOMEvent(): IDOMEventMeta;
    set onDOMEvent(value: IDOMEventMeta);
    get onWatchSetting(): IWatchSetting;
    set onWatchSetting(value: IWatchSetting);
    [key: string]: any,
}

type IEventHandler = (e: IMessage) => void;

export type IMessageMeta = { [key: string]: IEventHandler[] };

type IWatchFn = (newVal: any, oldVal: any, component: any) => void;
export type IWatchSetting = { [key: string]: IWatchFn[] };

export type IEventHandlersEventName = `on${keyof GlobalEventHandlersEventMap}`;
export type IDOMEventMeta = { [key in IEventHandlersEventName]: IEventHandler[] };



// interface IDOMEventMeta {
//     [key in IEventHandlersEventName]:string; 
//     // get [`on${keyof GlobalEventHandlersEventMap}`]:()=>void;
//     // set eventName(value: `on${keyof GlobalEventHandlersEventMap } `);
//     // get eventFunction(): (e: Event) => void;
//     // set eventFunction(value: (e: Event) => void);
// }

export interface ICustomSchema {
    eventSpecification: {
        inputEvent: IEventSpecificationEvent[],
        outputEvent: IEventSpecificationEvent[]
    }
    optionsView: {
        list: IOptionsView[],
    }
}
export interface IOptionsViewRules {
    required: boolean,
    message: string,
    trigger: string[]
}

export interface IOptionsView {
    type: string,
    label: string,
    options: {
        type: string,
        width: string,
        defaultValue: string,
        placeholder: string,
        clearable: boolean,
        maxLength: number,
        prepend: string,
        append: string,
        tooptip: string,
        hidden: boolean,
        disabled: boolean,
        dynamicHide: boolean,
        dynamicHideValue: string,
    },
    model: string,
    key: string,
    rules: IOptionsViewRules[],
}

export interface IEventSpecificationEvent {
    text: string,
    eventType: string,
    messageSchema: string,
    messageDemo: string,
}

export interface IMessage {
    header: {
        src: string,
        srcType: string,
        dst: string,
        dstType: string,
        fn?: (messagebody: IMessage) => any,
        reply?: {
            resolve: (messagebody: IMessage) => any,
            reject: (messagebody: IMessage) => any
        },
    },
    body: object
}