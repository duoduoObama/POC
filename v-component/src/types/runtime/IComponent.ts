import { IMessage } from "./IMessage";
import { ISchema } from "./IModelSchema";


export interface IComponent {
    model?: ISchema,
    onMessage(message: IMessage): void;
    sendMessage(message: IMessage): Promise<any>;
}


