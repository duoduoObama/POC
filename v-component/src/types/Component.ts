import { LitElement } from "lit";
import { EVENTBUS_NAME } from "../components/constent";
import { IComponent, IMessage, ISchema } from "./IComponent";

export class Component extends LitElement implements IComponent {

    model!: ISchema;
    eventObj: { [key: string]: Function } = {};
    constructor() {
        super();
    }

    /**
     * 添加监听器
     * @param eventName 事件名称
     * @param listener 
     */
    addListener(eventName: string, listener: (imessage: any) => void) {
        const method = function (message: any) {
            listener(message.detail);
        }
        this.eventObj[eventName] = method;
        this.addEventListener(eventName, method);
    }

    /**
     * 移除监听器
     * @param eventName 
     * @param listener 
     */
    removeListener(eventName: string) {
        this.removeEventListener(eventName, this.eventObj[eventName] as never);
    }

    /**
     * 获取监听的事件
     * @returns 
     */
    getListener() {
        return this.eventObj;
    }

    /**
     * 总线消息分发
     * @param imessage 
     */
    onMessage(imessage: IMessage): void {
        const { header } = imessage;
        const { dstType } = header;
        const customEvent = new CustomEvent(dstType, { detail: imessage });
        this.dispatchEvent(customEvent);
    }

    /**
     * 发送消息
     * @param imessage IMessage
     * @returns 
     */
    sendMessage(imessage: IMessage): Promise<any> {
        const reply = new Promise((resolve, reject) => {
            const { header } = imessage;
            Object.assign(header, { reply: { resolve, reject } });
            const customEvent = new CustomEvent(EVENTBUS_NAME, { detail: imessage });
            window.dispatchEvent(customEvent);
            console.log(`发送完毕，等待回复`);
        });

        return reply;
    }

    initModel(): void {
        // throw new Error("Method not implemented.");
    }
}