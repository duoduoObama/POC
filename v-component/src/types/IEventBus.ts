import { IComponent, IMessage } from "./IComponent";

export interface IEventBus {
    sendMessage(imessage: IMessage): Promise<any>;
    setDispatchEvent(dispatcher: (imessage: IMessage) => void): void;
}


export interface IconfigData {
    pageData: {

    },
    componentsArray: IComponent[],
    dynamicHTML: string,
}



