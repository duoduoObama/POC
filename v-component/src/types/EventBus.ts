import { EVENTBUS_NAME } from "../components/constent";
import { IMessage } from "./IComponent";
import { IEventBus } from "./IEventBus";

class EventBus implements IEventBus {

    /**
     * 事件
     */
    private eventObj: { [key: string]: Function } = {};

    /**
     * 发送消息
     * @param imessage 
     * @returns 
     */
    sendMessage(imessage: IMessage): Promise<any> {
        const reply = new Promise((resolve, reject) => {
            const { header } = imessage;
            Object.assign(header, { reply: { resolve, reject } });
            const customEvent = new CustomEvent(EVENTBUS_NAME, { detail: imessage });
            this.setDispatchEvent(customEvent);
            console.log(`发送完毕，等待回复`);
        });

        return reply;
    }

    /**
     * 消息分发
     * @param imessage 
     */
    setDispatchEvent(imessage: any): void {
        window.dispatchEvent(imessage);
    }

    /**
     * 接收消息
     * @param eventName 事件名称
     * @param callBack 
     */
    onMessage(eventName: string, callBack: Function): void {
        this.eventObj[eventName] = callBack;
        window.addEventListener(eventName, (event: any) => {
            callBack && callBack(event)
        });
    }

    /**
     * 获取总线事件
     * @returns 
     */
    getListener(): { [key: string]: Function } {
        return this.eventObj;
    }
}


export const eventBus = new EventBus();