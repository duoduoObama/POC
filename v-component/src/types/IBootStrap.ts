import { IconfigData } from "./IEventBus";

/**
 * 启动函数接口
 */
export interface IBootStrap {
    bootStrap: (root: HTMLElement | string, config: IconfigData) => any;
}