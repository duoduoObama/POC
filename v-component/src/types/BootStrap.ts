import { IBootStrap } from "./IBootStrap";
import { IconfigData } from "./IEventBus";

declare global {
    const config: IconfigData;
}

class BootStrap implements IBootStrap {

    bootStrap(config: IconfigData): any {
        if (!config) {
            return;
        }
        const { pageData = {}, componentsArray = [] } = config;


        componentsArray.forEach(component => {
            const { componentName, initStyle, options, id } = component;
            const dragElement = document.createElement("div");
            const componentElement = document.createElement(componentName as string);

            dragElement.style.cssText = initStyle as string;
            componentElement.id = id as string;
            componentElement.dataset.data = JSON.stringify(options);
            componentElement.dataset.component = "true";
            dragElement.appendChild(componentElement);
            document.body.appendChild(dragElement);
        });
        this.pageModelBootStrap();
    }

    pageModelBootStrap() {
        const pageModel = new Proxy(config, {}) as IconfigData;

        pageModel.componentsArray.forEach((component) => {
            console.log(component);
            const { id, model = { eventSpecification: {} } } = component;
            const componentElement = document.getElementById(id as string) as any;
            const dragElement = componentElement.parentElement;
            if (componentElement.model) {
                component.model = new Proxy(componentElement.model, {}) as any;
            }
            component.initStyle = new Proxy(dragElement.style, {}) as any;
        })

        console.log(pageModel);
        window.pageModel = pageModel;
    }
}


export default new BootStrap();