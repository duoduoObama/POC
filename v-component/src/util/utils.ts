import { cloneDeep, isFunction, isObject } from "lodash-es";
import { ISchema, IWatchSetting } from "../types/IComponent";

/**
 * 生成hashid
 * @param {*} hashLength
 */
export function createHash(hashLength = 12) {
    // 默认长度 24
    return Array.from(Array(Number(hashLength) || 24), () => Math.floor(Math.random() * 36).toString(36)).join("");
}

/**
 * 生成onlyid
 * @param {*} hashLength
 * @param {*} key
 */
export function createHashId(hashLength = 12, key = `drag-`) {
    let probeId = `${key}${createHash(hashLength)}`;
    while (document.querySelectorAll(`#${probeId}`).length) {
        // dom存在即重新构成
        probeId = `${key}${createHash(hashLength)}`;
    }
    return probeId;
}


/**
 * 获取元素树
 * @param target 
 * @param param1 
 * @returns 
 */
export function getElementTree(target = document, { isEle = false } = {}) {
    if (!target || target.nodeName === "#text") return [];
    return [...target.querySelectorAll("[data-component]")].map((c: any) => {
        let { data, parentId } = c.dataset;
        const { x, y } = c.parentElement.dataset;
        const $el = c.componentInstance ? c.componentInstance.$el : null;
        data = JSON.parse(data);
        data.id = c.id;
        data.proxyInfo = c.data;
        data.parentId = parentId;
        data.style = c.parentElement.style.cssText;
        data.parentClass = c.parentElement.classList.value;
        data.x = x;
        data.y = y;
        data.children = $el ? getElementTree($el, { isEle }) : [];
        if (isEle) {
            data.$el = $el;
            data.componentInstance = c.componentInstance;
            data.shadowRoot = c.shadowRoot;
            data.srcElement = c;
            data.parentElement = c.parentElement;
        }
        return data;
    });
}

/**
 * 获取目标元素树
 * @param target 
 * @returns 
 */
export function getTargetElement(target = "") {
    if (!target) return null;

    const allElement = getElementTree(document, { isEle: true });

    const findTarget = (arr = [], targetId: string) => {
        let len = 0;
        let isBreak = true;
        let targetNode = null;
        while (arr[len] && isBreak) {
            const { id, children = [] } = arr[len];
            if (targetId === id) {
                targetNode = arr[len];
                isBreak = false;
                return targetNode;
            } else if (children && children.length) {
                const target = findTarget(children, targetId) as object;
                if (target) {
                    return target;
                }
            }
            len++;
        }
        return targetNode;
    };

    return findTarget(allElement as never[], target);
}

/**
 * 深度监听
 * @param obj 代理对象
 * @param cb 回调
 * @returns 
 */
export function deepWatchModelProxy(obj: ISchema) {

    if (isObject(obj) && !isFunction(obj)) {

        for (let key of Object.keys(obj)) {
            if (isObject(obj[key]) && !isFunction(obj[key])) {
                try {
                    obj[key] = deepWatchModelProxy(obj[key]) as any;
                } catch (error) {
                    // console.log(error);
                }

            }
        }

    }

    return new Proxy(obj, {

        /**
         * @param {Object, Array} target 设置值的对象
         * @param {String} key 属性
         * @param {any} value 值
         * @param {Object} receiver this
         */
        set: function (target: any, key, value: { [key: string]: {} }, receiver) {

            const copyValue = cloneDeep(target[key]);

            if (isObject(value) && !isFunction(obj)) {
                value = deepWatchModelProxy(value as any);
            }

            const onWatchFn = obj.onWatchSetting as IWatchSetting;
            const propertyWatch = onWatchFn[key as string] ?? [];
            propertyWatch.forEach((callBack: Function) => {
                callBack(value, copyValue, obj);
            });
            return Reflect.set(target, key, value, receiver);

        },
        deleteProperty(target, key) {
            return Reflect.deleteProperty(target, key);
        }

    });
} 