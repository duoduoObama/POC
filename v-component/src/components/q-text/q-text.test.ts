import { expect, describe, test } from "vitest";
import { IMessage, ISchema, Ischemas } from "../../types/IComponent";
import { deepWatchModelProxy } from "../../util/utils";
import { IQtextOptions } from "./IQText";
import { JSDOM } from "jsdom";
import { cloneDeep } from "lodash-es";

// render函数测试 是否获取data数据
const data: IQtextOptions = { text: "文本测试" };
const render = () => {
  const { text } = data;
  return text;
};
describe.skip("测试render函数", () => {
  test("对render函数进行测试", () => {
    expect(render()).toEqual("文本测试");
  });
});

// initModel函数测试 深度监听
let model!: ISchema;
const initModel = function (this: any) {
  model = deepWatchModelProxy({
    id: "",
    componentName: "q-text",
    type: "文本",
    text: "文本",
    group: ["文本"],
    createTime: "2022-07-14T05:48:36.830Z",
    image: "",
    _initStyle: "",
    initStyle: "",
    description: "文本组件,可以编写文字信息",
    options: {
      text: "文本输入框测试",
    },
    schema: {
      eventSpecification: {
        inputEvent: [
          {
            text: "更改组件数据",
            eventType: "changeInfo",
            messageSchema: "",
            messageDemo: "",
          },
        ],
        outputEvent: [
          {
            text: "组件点击数据",
            eventType: "change",
            messageSchema: "",
            messageDemo: "文本数据1",
          },
        ],
      },
      optionsView: {
        list: [
          {
            type: "input",
            label: "输入框",
            options: {
              type: "text",
              width: "100%",
              defaultValue: "",
              placeholder: "请输入",
              clearable: false,
              maxLength: 0,
              prepend: "",
              append: "",
              tooptip: "",
              hidden: false,
              disabled: false,
              dynamicHide: false,
              dynamicHideValue: "",
            },
            model: "text",
            key: "text",
            rules: [
              {
                required: false,
                message: "必填项",
                trigger: ["blur"],
              },
            ],
          },
        ],
      },
    },
    _eventSpecification: {
      inputEvent: [
        {
          text: "更改组件数据",
          eventType: "changeInfo",
          messageSchema: "",
          messageDemo: "",
        },
      ],
      inputCustomEvent: [
        {
          text: "更改组件数据",
          eventType: "changeInfo",
          messageSchema: "",
          messageDemo: "",
        },
      ],
      outputEvent: [
        {
          text: "组件点击数据",
          eventType: "click",
          messageSchema: "",
          messageDemo: "文本数据1",
        },
      ],
    },
    eventSpecification: {
      inputEvent: [
        {
          text: "更改组件数据",
          eventType: "changeInfo",
          messageSchema: "",
          messageDemo: "",
        },
      ],
      inputCustomEvent: [
        {
          text: "更改组件数据",
          eventType: "changeInfo",
          messageSchema: "",
          messageDemo: "",
        },
      ],
      outputEvent: [
        {
          text: "组件点击数据",
          eventType: "click",
          messageSchema: "",
          messageDemo: "文本数据1",
        },
      ],
    },
    _onMessageMeta: {},
    _onDOMEvent: {},
    _onWatchSetting: {},
    onMessageMeta: {},
    onDOMEvent: {},
    onWatchSetting: {},
  });
  return model;
};
describe.skip("测试Model初始化", () => {
  test("对initModel函数进行测试", () => {
    initModel();
    Object.keys(Ischemas).forEach((item) => {
      expect(model).toHaveProperty(item);
    });
  });
});

// clickFont函数测试 是否发送消息
const dom = new JSDOM(`<p>测试文本</p>`);
let messages: any = null;
dom.window.document.querySelector("p")?.addEventListener("click", function (e) {
  return onSendMessage(e, data, "text");
});

dom.window.document.querySelector("p")?.click();
function onSendMessage(e: Event, node: any, index: number | string) {
  const message: IMessage = {
    header: {
      src: "",
      dst: "",
      srcType: e.type,
      dstType: "",
    },
    body: cloneDeep(data),
  };
  messages = message;
  return message;
}

// 进行click的事件监听
describe.skip("测试dom点击事件传递Messge", () => {
  test("对clickFont dom点击事件进行测试", () => {
    dom.window.document.querySelector("p")?.click();

    expect(messages).toEqual({
      header: { src: "", dst: "", srcType: "click", dstType: "" },
      body: { text: "文本测试" },
    });
  });
});

// 测试类型
describe.skip("测试model是否属于", () => {
  test("对属性测试进行", () => {
    const obj = {
      componentName: "曹義伟",
      id: 123,
      type: "123",
      text: "123",
    };
    // 用于判断是否满足key列表
    let flag = true;

    Object.keys(Ischemas).length
      ? Object.keys(Ischemas).forEach((item) => {
          if (!Object.keys(obj).includes(item)) flag = false;
        })
      : null;
    expect(flag).toEqual(true);
  });
});

// 判断属性
const johnInvoice = {
  isActive: true,
  customer: {
    first_name: "John",
    last_name: "Doe",
    location: "China",
  },
  total_amount: 5000,
  items: [
    {
      type: "apples",
      quantity: 10,
    },
    {
      type: "oranges",
      quantity: 5,
    },
  ],
};

const obj = {
  componentName: "曹義伟",
  id: 123,
  type: "123",
  text: "123",
  group: "string",
  createTime: "Date",
  image: "string",
  initStyle: "string",
  description: "string",
  options: "any",
  schema: "ICustomSchema",
  onMessageMeta: "IMessageMeta",
  onDOMEvent: "IDOMEventMeta",
  onWatchSetting: "IWatchSetting",
};
test.skip("测试是否满足属性设置", () => {
  Object.keys(Ischemas).forEach((item) => {
    expect(obj).toHaveProperty(item);
  });
});
