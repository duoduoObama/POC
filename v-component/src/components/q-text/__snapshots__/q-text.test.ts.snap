// Vitest Snapshot v1

exports[`foo snapshot test 1`] = `
{
  "foo": {
    "x": 1,
    "y": 2,
  },
}
`;
